### PR TITLE
MHA: Fix regression and apply bias flag to both in/out proj

### LINKED ISSUE #52257: MultiHeadAttention bias=False regression
## 🐛 Bug

`MultiHeadAttention` does not respect `bias=False` in the output projection. It seems that it was changed as part of a PR: https://github.com/pytorch/pytorch/commit/eace0533985641d9c2f36e43e3de694aca886bd9#diff-61bd5e4f390b965228ead6bb0efcad5b9b5299d833bcf5c2aa11ebaebb39ba29L794

Is this the intended behavior?

@ezyang 

## To Reproduce

1. Create a MultiHeadAttention module with `torch.nn.MultiheadAttention(N, H, bias=False)`
1. The module's `out_proj` field as a non-None, nonzero bias

## Expected behavior

The bias in `out_proj` should be None.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.1 and latest master
 - OS (e.g., Linux): All
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

## Additional context

N/A

cc @ezyang @gchanan @zhangguanheng66

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #52257

## Background
Reverts MHA behavior for `bias` flag to that of v1.5: flag enables or disables both in and out projection biases.

Updates type annotations for both in and out projections biases from `Tensor` to `Optional[Tensor]` for `torch.jit.script` usage.

Note: With this change, `_LinearWithBias` defined in `torch/nn/modules/linear.py` is no longer utilized. Completely removing it would require updates to quantization logic in the following files:
```
test/quantization/test_quantized_module.py
torch/nn/quantizable/modules/activation.py
torch/nn/quantized/dynamic/modules/linear.py
torch/nn/quantized/modules/linear.py
torch/quantization/quantization_mappings.py
```
This PR takes a conservative initial approach and leaves these files unchanged.

**Is it safe to fully remove `_LinearWithBias`?** 

## Test Plan
```
python test/test_nn.py TestNN.test_multihead_attn_no_bias
```

## BC-Breaking Note
In v1.6, the behavior of `MultiheadAttention`'s `bias` flag was incorrectly changed to affect only the in projection layer. That is, setting `bias=False` would fail to disable the bias for the out projection layer. This regression has been fixed, and the `bias` flag now correctly applies to both the in and out projection layers.

### CODE REVIEW

**Comment by gchanan** [comment]
I don't think we need to worry about removing _LinearWithBias.

**Review by gchanan — APPROVED** [review]

**Review by zhangguanheng66 — APPROVED** [review]

File: torch/nn/quantizable/modules/activation.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def __init__(self, embed_dim: int, num_heads: int,
         # TODO: The use of the `_LinearWithBias` increases the quantization noise
```
Comment (gchanan) [inline torch/nn/quantizable/modules/activation.py:65]: is this comment still relevant?

**Review by gchanan — APPROVED** [review]
lgmt.  @vkuzo mind taking a look at the quantization code?

File: torch/nn/quantizable/modules/activation.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/quantizable/modules/activation.py:65]: I went back and forth on this, but I left it in for the case where an older model is loaded and then quantized. Is this valid? The comment should probably be at least updated to reflect that `out_proj` could or could not be a `_LinearWithBias`.

**Comment by vkuzo** [comment]
cc @z-a-f  as an fyi

**Pull request closed without merging**